lib/ecto/changeset.ex

```diff
@@ -1015,7 +1024,22 @@ defmodule Ecto.Changeset do
           "please use put_#{tag}/4 instead"
   end
 
-  defp put_change(data, changes, errors, valid?, key, value, _type) do
+  defp put_change(data, changes, errors, valid?, key, value, nil) do
+    put_change(data, changes, errors, valid?, key, value)
+  end
+
+  defp put_change(data, changes, errors, valid?, key, value, type) do
+    case Ecto.Type.cast(type, value) do
+      {:ok, ^value} ->
+        put_change(data, changes, errors, valid?, key, value)
+      {:ok, _} ->
+        raise Ecto.CastError, type: type, value: value
```

Comment:
as we discussed e.g. `"10"` can be put for `:integer` field. Which makes me think that `cannot cast "10" as :integer` can be a confusing error message, because casting in other contexts would allow that. Should we do e.g.:

```elixir
raise Ecto.CastError, type: type, value: value, message: "value #{inspect value} must exactly match #{inspect type}"
```

or something similar? not sure what's the best wording here.